File: apex_dashboard/apex_dashboard/page/suppliers_dashboard/suppliers_dashboard.py
import frappe
from frappe import _
from frappe.utils import flt, getdate, add_months, add_days, get_first_day, get_last_day, today
import requests

def get_exchange_rates():
	"""
	Fetches exchange rates from OpenExchangeRates API (same as Liquidity Dashboard).
	"""
	# Get API key from site config
	api_key = frappe.conf.get("openexchangerates_api_key")
	
	# Check cache first
	cache_key = "suppliers_dashboard_rates"
	cached_rates = frappe.cache().get_value(cache_key)
	if cached_rates:
		return cached_rates
	
	if not api_key:
		return get_fallback_rates()
	
	try:
		url = f"https://openexchangerates.org/api/latest.json?app_id={api_key}&base=USD"
		response = requests.get(url, timeout=5)
		response.raise_for_status()
		data = response.json()
		
		# Convert to EGP base
		if "EGP" not in data["rates"]:
			return get_fallback_rates()
		
		usd_to_egp = data["rates"]["EGP"]
		rates = {'EGP': 1.0}
		
		for currency in ['USD', 'EUR', 'SAR']:
			if currency in data["rates"]:
				rate_in_usd = data["rates"][currency]
				if currency == "USD":
					rates[currency] = usd_to_egp
				else:
					# Cross-currency conversion: EUR to EGP = (USD to EGP) × (EUR to USD)
					if rate_in_usd > 0:
						rates[currency] = usd_to_egp / rate_in_usd
					else:
						rates[currency] = 0.0
		
		# Cache for 1 hour
		frappe.cache().set_value(cache_key, rates, expires_in_sec=3600)
		return rates
	except Exception as e:
		# Log error but don't fail if log fails (e.g., error message too long)
		try:
			error_msg = str(e)[:100]  # Truncate to 100 chars to avoid "Title too long" error
			frappe.log_error(f"OpenExchangeRates API Error: {error_msg}", "Exchange Rates Error")
		except:
			pass  # Silently ignore log errors
		return get_fallback_rates()

def get_fallback_rates():
	"""Fallback to Currency Exchange table."""
	rates = {'EGP': 1.0}
	for curr in ['USD', 'EUR', 'SAR']:
		# Get latest exchange rate
		rate_data = frappe.db.sql("""
			SELECT exchange_rate 
			FROM `tabCurrency Exchange`
			WHERE from_currency = %s 
				AND to_currency = 'EGP'
				AND for_selling = 1
			ORDER BY date DESC
			LIMIT 1
		""", curr, as_dict=1)
		
		if rate_data and len(rate_data) > 0:
			rates[curr] = flt(rate_data[0].get('exchange_rate'))
		else:
			# Hard defaults
			if curr == 'USD':
				rates[curr] = 50.0
			elif curr == 'EUR':
				rates[curr] = 55.0
			elif curr == 'SAR':
				rates[curr] = 13.0
	return rates

def get_period_dates(period):
	current_date = getdate(today())
	
	if period == "Today":
		return current_date, current_date
	elif period == "This Week":
		start = add_days(current_date, -current_date.weekday())
		end = add_days(start, 6)
		return start, end
	elif period == "This Month":
		return get_first_day(current_date), get_last_day(current_date)
	elif period == "Last Month":
		last_month = add_months(current_date, -1)
		return get_first_day(last_month), get_last_day(last_month)
	elif period == "This Year":
		return getdate(f"{current_date.year}-01-01"), getdate(f"{current_date.year}-12-31")
	elif period == "Last Year":
		last_year = current_date.year - 1
		return getdate(f"{last_year}-01-01"), getdate(f"{last_year}-12-31")
	elif period == "All Time":
		return getdate("2000-01-01"), current_date
	else:
		return get_first_day(current_date), get_last_day(current_date)

@frappe.whitelist()
def get_dashboard_data(company=None, period="All Time", from_date=None, to_date=None, fiscal_year=None):
	"""Get comprehensive supplier dashboard data"""
	if not company:
		company = frappe.defaults.get_user_default("Company")
	
	# Determine Date Range
	if fiscal_year:
		from_date, to_date = frappe.db.get_value("Fiscal Year", fiscal_year, ["year_start_date", "year_end_date"])
	elif period == "Custom" and from_date and to_date:
		from_date = getdate(from_date)
		to_date = getdate(to_date)
	else:
		if not period:
			period = "All Time"
		from_date, to_date = get_period_dates(period)
	
	currency = frappe.get_value("Company", company, "default_currency") or "EGP"
	
	# Get exchange rates
	exchange_rates = get_exchange_rates()
	
	# 1. Total Payables (Outstanding Purchase Invoices) - Grouped by Currency
	# Filter by posting_date to show outstanding invoices FROM that period
	payables_data = frappe.db.sql("""
		SELECT 
			pi.supplier,
			s.supplier_name,
			s.supplier_group,
			pi.currency,
			COUNT(DISTINCT pi.name) as invoice_count,
			SUM(pi.outstanding_amount) as outstanding,
			SUM(pi.grand_total) as total_amount
		FROM `tabPurchase Invoice` pi
		JOIN `tabSupplier` s ON pi.supplier = s.name
		WHERE pi.docstatus = 1
			AND pi.company = %s
			AND pi.outstanding_amount > 0
			AND s.disabled = 0
			AND pi.posting_date BETWEEN %s AND %s
		GROUP BY pi.supplier, pi.currency
		ORDER BY pi.currency, outstanding DESC
	""", (company, from_date, to_date), as_dict=1)
	
	# Group by currency
	payables_by_currency = {}
	total_payables_egp = 0  # Total in EGP
	payables_breakdown = {}  # Breakdown by currency
	
	for d in payables_data:
		curr = d.get('currency') or 'EGP'
		if curr not in payables_by_currency:
			payables_by_currency[curr] = {
				'total': 0,
				'count': 0,
				'details': []
			}
		payables_by_currency[curr]['total'] += d.get('outstanding', 0)
		payables_by_currency[curr]['count'] += d.get('invoice_count', 0)
		payables_by_currency[curr]['details'].append(d)
		
		# Add to breakdown
		if curr not in payables_breakdown:
			payables_breakdown[curr] = 0
		payables_breakdown[curr] += d.get('outstanding', 0)
		
		# Add to total in EGP
		total_payables_egp += d.get('outstanding', 0) * exchange_rates.get(curr, 1.0)
	
	# Calculate Total Purchase Volume (all invoices, not just outstanding)
	total_purchase_data = frappe.db.sql("""
		SELECT 
			pi.currency,
			SUM(pi.grand_total) as total
		FROM `tabPurchase Invoice` pi
		WHERE pi.docstatus = 1
			AND pi.company = %s
			AND pi.posting_date BETWEEN %s AND %s
		GROUP BY pi.currency
	""", (company, from_date, to_date), as_dict=1)
	
	total_purchase_egp = 0
	purchase_breakdown = {}  # Breakdown by currency
	
	for d in total_purchase_data:
		curr = d.get('currency') or 'EGP'
		purchase_breakdown[curr] = d.get('total', 0)
		total_purchase_egp += d.get('total', 0) * exchange_rates.get(curr, 1.0)
	
	# 2. Total Paid (Payment Entries)
	paid_data = frappe.db.sql("""
		SELECT 
			pe.party as supplier,
			s.supplier_name,
			COUNT(DISTINCT pe.name) as payment_count,
			SUM(pe.paid_amount) as paid_amount
		FROM `tabPayment Entry` pe
		JOIN `tabSupplier` s ON pe.party = s.name
		WHERE pe.docstatus = 1
			AND pe.company = %s
			AND pe.party_type = 'Supplier'
			AND pe.payment_type = 'Pay'
			AND pe.posting_date BETWEEN %s AND %s
			AND s.disabled = 0
		GROUP BY pe.party
		ORDER BY paid_amount DESC
	""", (company, from_date, to_date), as_dict=1)
	
	total_paid = sum(d.get('paid_amount', 0) for d in paid_data)
	paid_count = sum(d.get('payment_count', 0) for d in paid_data)
	
	# 3. Active Suppliers (with transactions in period)
	active_suppliers = frappe.db.sql("""
		SELECT DISTINCT pi.supplier, s.supplier_name, s.supplier_group
		FROM `tabPurchase Invoice` pi
		JOIN `tabSupplier` s ON pi.supplier = s.name
		WHERE pi.docstatus = 1
			AND pi.company = %s
			AND pi.posting_date BETWEEN %s AND %s
			AND s.disabled = 0
		ORDER BY s.supplier_name
	""", (company, from_date, to_date), as_dict=1)
	
	# Group by supplier group
	supplier_groups = {}
	for sup in active_suppliers:
		group = sup.get('supplier_group') or 'Other'
		if group not in supplier_groups:
			supplier_groups[group] = []
		supplier_groups[group].append({
			'name': sup.get('supplier_name'),
			'code': sup.get('supplier')
		})
	
	# 4. Overdue Payments - Grouped by Currency
	overdue_data = frappe.db.sql("""
		SELECT 
			pi.supplier,
			s.supplier_name,
			pi.currency,
			COUNT(DISTINCT pi.name) as overdue_count,
			SUM(pi.outstanding_amount) as overdue_amount,
			MAX(DATEDIFF(CURDATE(), pi.due_date)) as days_overdue
		FROM `tabPurchase Invoice` pi
		JOIN `tabSupplier` s ON pi.supplier = s.name
		WHERE pi.docstatus = 1
			AND pi.company = %s
			AND pi.outstanding_amount > 0
			AND pi.due_date < CURDATE()
			AND pi.posting_date BETWEEN %s AND %s
			AND s.disabled = 0
		GROUP BY pi.supplier, pi.currency
		ORDER BY pi.currency, overdue_amount DESC
	""", (company, from_date, to_date), as_dict=1)
	
	# Group by currency
	overdue_by_currency = {}
	for d in overdue_data:
		curr = d.get('currency') or 'EGP'
		if curr not in overdue_by_currency:
			overdue_by_currency[curr] = {
				'total': 0,
				'count': 0,
				'details': []
			}
		overdue_by_currency[curr]['total'] += d.get('overdue_amount', 0)
		overdue_by_currency[curr]['count'] += d.get('overdue_count', 0)
		overdue_by_currency[curr]['details'].append(d)
	
	# 5. Top Suppliers by Total Purchase Volume - With Currency and EGP Conversion
	# Build CASE statement for exchange rates
	rate_cases = []
	for curr, rate in exchange_rates.items():
		rate_cases.append(f"WHEN pi.currency = '{curr}' THEN {rate}")
	
	rate_case_sql = f"CASE {' '.join(rate_cases)} ELSE 1.0 END"
	
	top_suppliers = frappe.db.sql(f"""
		SELECT 
			pi.supplier,
			s.supplier_name,
			pi.currency,
			COUNT(DISTINCT pi.name) as invoice_count,
			SUM(pi.grand_total) as total_purchase,
			SUM(pi.outstanding_amount) as outstanding,
			SUM(pi.grand_total * {rate_case_sql}) as total_purchase_egp,
			MIN(pi.posting_date) as first_invoice_date,
			MAX(pi.posting_date) as last_invoice_date
		FROM `tabPurchase Invoice` pi
		JOIN `tabSupplier` s ON pi.supplier = s.name
		WHERE pi.docstatus = 1
			AND pi.company = %s
			AND s.disabled = 0
			AND pi.posting_date BETWEEN %s AND %s
		GROUP BY pi.supplier, pi.currency
		ORDER BY total_purchase_egp DESC
		LIMIT 20
	""", (company, from_date, to_date), as_dict=1)
	
	return {
		'currency': currency,
		'exchange_rates': exchange_rates,
		'total_payables_egp': total_payables_egp,
		'payables_breakdown': payables_breakdown,
		'total_purchase_egp': total_purchase_egp,
		'purchase_breakdown': purchase_breakdown,
		'payables_by_currency': payables_by_currency,
		'paid': {
			'total': total_paid,
			'count': paid_count,
			'details': paid_data[:10],  # Top 10
			'period': f"{from_date.strftime('%d %b')} - {to_date.strftime('%d %b %Y')}"
		},
		'active_suppliers': {
			'total': len(active_suppliers),
			'groups': supplier_groups
		},
		'overdue_by_currency': overdue_by_currency,
		'top_suppliers': top_suppliers
	}

File: apex_dashboard/apex_dashboard/page/suppliers_dashboard/suppliers_dashboard.js
frappe.pages['suppliers_dashboard'].on_page_load = function (wrapper) {
	new SuppliersDashboard(wrapper);
};

class SuppliersDashboard {
	constructor(wrapper) {
		this.wrapper = $(wrapper);

		this.page = frappe.ui.make_app_page({
			parent: wrapper,
			title: 'Suppliers Dashboard',
			single_column: true
		});

		// Load CSS
		frappe.require('/assets/apex_dashboard/page/suppliers_dashboard/suppliers_dashboard.css');

		const TEMPLATE = `
		<div class="suppliers-dashboard-container dashboard-template">
			<div class="dashboard-header">
				<div class="header-content">
					<button class="btn-glass" id="back-btn">
						<i class="fa fa-arrow-left"></i> Hub
					</button>
					<div class="header-text">
						<h1>Suppliers Overview</h1>
						<p class="subtitle">Track your supplier performance and payables</p>
					</div>
				</div>
				<div class="header-actions">
					<div class="total-card" style="padding: 10px 20px; min-width: 200px;">
						<span class="label" style="font-size: 12px; color: #9ca3af; display: block;">Total Payables</span>
						<span class="value" id="header-total-payables" style="font-size: 20px; font-weight: bold;">Loading...</span>
					</div>
					<button class="btn-glass" id="refresh-btn">
						<i class="fa fa-refresh"></i>
					</button>
				</div>
			</div>

			<div id="dashboard-loader" class="dashboard-loader">
				<div class="spinner"></div>
				<p>Loading dashboard...</p>
			</div>

			<div id="dashboard-content" style="display: none;">
				<!-- Content will be injected here -->
			</div>
		</div>
		`;

		// Load HTML Template
		$(TEMPLATE).appendTo(this.page.body);

		// Hide standard page title since we have a custom header
		if (this.page.page_header) {
			this.page.page_header.hide();
		} else {
			this.wrapper.find('.page-head').hide();
		}

		this.format_currency = this.format_currency.bind(this);
		this.setup_filters();
		this.bind_events();
		this.load_data();
	}

	bind_events() {
		this.wrapper.find('#back-btn').on('click', () => {
			frappe.set_route('apex_dashboards');
		});

		this.wrapper.find('#refresh-btn').on('click', () => {
			this.load_data();
		});
	}

	setup_filters() {
		// Company filter
		this.page.add_field({
			fieldname: 'company',
			label: __('Company'),
			fieldtype: 'Link',
			options: 'Company',
			default: frappe.defaults.get_user_default('Company'),
			change: () => this.load_data()
		});

		// Period filter
		this.page.add_field({
			fieldname: 'period',
			label: __('Period'),
			fieldtype: 'Select',
			options: ['', 'Today', 'This Week', 'This Month', 'Last Month', 'This Year', 'Last Year', 'Fiscal Year', 'Custom', 'All Time'],
			default: 'All Time',
			change: () => {
				const period = this.page.fields_dict.period.get_value();

				// Toggle Date Fields
				const isCustom = period === 'Custom';
				if (isCustom) {
					this.page.fields_dict.from_date.$wrapper.removeClass('hide').show();
					this.page.fields_dict.to_date.$wrapper.removeClass('hide').show();
				} else {
					this.page.fields_dict.from_date.$wrapper.addClass('hide').hide();
					this.page.fields_dict.to_date.$wrapper.addClass('hide').hide();
				}

				// Toggle Fiscal Year Field
				const isFiscalYear = period === 'Fiscal Year';
				if (isFiscalYear) {
					this.page.fields_dict.fiscal_year.$wrapper.removeClass('hide').show();
				} else {
					this.page.fields_dict.fiscal_year.$wrapper.addClass('hide').hide();
					this.page.fields_dict.fiscal_year.set_input(''); // Clear value when hidden
				}

				if (period !== 'Custom' && period !== 'Fiscal Year') {
					this.load_data();
				}
			}
		});

		// Fiscal Year Filter (Hidden by default)
		this.page.add_field({
			fieldname: 'fiscal_year',
			label: __('Select Year'),
			fieldtype: 'Link',
			options: 'Fiscal Year',
			change: () => {
				const fiscal_year = this.page.fields_dict.fiscal_year.get_value();
				if (fiscal_year) {
					this.load_data();
				}
			}
		});

		// Date range filters (hidden by default)
		this.page.add_field({
			fieldname: 'from_date',
			label: __('From Date'),
			fieldtype: 'Date',
			change: () => this.load_data()
		});

		this.page.add_field({
			fieldname: 'to_date',
			label: __('To Date'),
			fieldtype: 'Date',
			change: () => this.load_data()
		});

		// Initialize visibility based on default period
		this.page.fields_dict.period.$input.trigger('change');
	}

	load_data() {
		const company = this.page.fields_dict.company.get_value();
		const period = this.page.fields_dict.period.get_value();
		const from_date = this.page.fields_dict.from_date.get_value();
		const to_date = this.page.fields_dict.to_date.get_value();
		const fiscal_year = this.page.fields_dict.fiscal_year.get_value();

		this.show_loader(true);

		frappe.call({
			method: 'apex_dashboard.apex_dashboard.page.suppliers_dashboard.suppliers_dashboard.get_dashboard_data',
			args: {
				company: company,
				period: period,
				from_date: from_date,
				to_date: to_date,
				fiscal_year: fiscal_year
			},
			callback: (r) => {
				this.show_loader(false);
				if (r.message) {
					this.render(r.message);
				} else {
					this.wrapper.find('#dashboard-content').html('<div class="no-data">No data available</div>').show();
				}
			},
			error: (r) => {
				this.show_loader(false);
				frappe.msgprint(__('Failed to load dashboard data'));
			}
		});
	}

	show_loader(show) {
		if (show) {
			this.wrapper.find('#dashboard-loader').show();
			this.wrapper.find('#dashboard-content').hide();
		} else {
			this.wrapper.find('#dashboard-loader').hide();
			this.wrapper.find('#dashboard-content').show();
		}
	}

	render(data) {
		const currency = data.currency || 'EGP';

		// Update Header Total
		this.wrapper.find('#header-total-payables').text(this.format_currency(data.total_payables_egp, 'EGP'));

		// Render payables cards for each currency
		let payablesCards = '';
		if (data.payables_by_currency) {
			for (const [curr, payablesData] of Object.entries(data.payables_by_currency)) {
				payablesCards += this.render_card(`💰 Payables (${curr})`, payablesData, curr, '#e74c3c', 'payables');
			}
		}

		// Render overdue cards for each currency
		let overdueCards = '';
		if (data.overdue_by_currency) {
			for (const [curr, overdueData] of Object.entries(data.overdue_by_currency)) {
				overdueCards += this.render_card(`⚠️ Overdue (${curr})`, overdueData, curr, '#f39c12', 'overdue');
			}
		}

		const html = `
			<div class="suppliers-dashboard">
				<!-- Total Summary Cards -->
				<div class="total-summary-cards">
					${this.render_total_card('💰 Total Payables', data.total_payables_egp, 'EGP', '#e74c3c', data.exchange_rates, data.payables_breakdown)}
					${this.render_total_card('📊 Total Purchase Volume', data.total_purchase_egp, 'EGP', '#667eea', data.exchange_rates, data.purchase_breakdown)}
				</div>
				
				<!-- Summary Cards -->
				<div class="summary-cards">
					${payablesCards}
					${this.render_card('💵 Total Paid', data.paid, currency, '#27ae60', 'paid')}
					${this.render_card('📦 Active Suppliers', data.active_suppliers, currency, '#3498db', 'active')}
					${overdueCards}
				</div>
				
				<!-- Top Suppliers Section -->
				<div class="top-suppliers-section">
					<h3>📊 Top Suppliers by Purchase Volume</h3>
					${this.render_top_suppliers(data.top_suppliers, data.exchange_rates)}
				</div>
			</div>
		`;

		this.wrapper.find('#dashboard-content').html(html);
	}

	render_card(title, data, currency, color, type) {
		let detailsHtml = '';

		if (type === 'payables' && data.details && data.details.length > 0) {
			detailsHtml = `
				<div class="card-details">
					${data.details.map(item => `
						<div class="card-detail-item clickable" onclick="frappe.set_route('List', 'Purchase Invoice', {'supplier': '${item.supplier}', 'outstanding_amount': ['>', 0]})" title="View outstanding invoices">
							<span class="detail-name">${item.supplier_name}</span>
							<div class="detail-stats">
								<span class="detail-qty amount-red">${this.format_currency(item.outstanding, currency)}</span>
								<span class="detail-types">${item.invoice_count} invoices</span>
							</div>
						</div>
					`).join('')}
				</div>
			`;
		} else if (type === 'paid' && data.details && data.details.length > 0) {
			detailsHtml = `
				<div class="period-info">${data.period || ''}</div>
				<div class="card-details">
					${data.details.map(item => `
						<div class="card-detail-item clickable" onclick="frappe.set_route('List', 'Payment Entry', {'party': '${item.supplier}'})" title="View payments">
							<span class="detail-name">${item.supplier_name}</span>
							<div class="detail-stats">
								<span class="detail-qty amount-green">${this.format_currency(item.paid_amount, currency)}</span>
								<span class="detail-types">${item.payment_count} payments</span>
							</div>
						</div>
					`).join('')}
				</div>
			`;
		} else if (type === 'active' && data.groups && data.groups.length > 0) {
			detailsHtml = `
				<div class="card-details">
					${data.groups.map(group => `
						<div class="supplier-group">
							<span class="group-name">${group.name}</span>
							<span class="group-count">${group.count} suppliers</span>
						</div>
					`).join('')}
				</div>
			`;
		} else if (type === 'overdue' && data.details && data.details.length > 0) {
			detailsHtml = `
				<div class="card-details">
					${data.details.map(item => `
						<div class="card-detail-item clickable" onclick="frappe.set_route('List', 'Purchase Invoice', {'supplier': '${item.supplier}', 'outstanding_amount': ['>', 0], 'due_date': ['<', frappe.datetime.get_today()]})" title="View overdue invoices">
							<span class="detail-name">${item.supplier_name}</span>
							<div class="detail-stats">
								<span class="detail-qty amount-orange">${this.format_currency(item.overdue_amount, currency)}</span>
								<span class="detail-types">${item.days_overdue} days overdue</span>
							</div>
						</div>
					`).join('')}
				</div>
			`;
		}

		const value = type === 'active' ? data.total : this.format_currency(data.total, currency);
		const subtitle = type === 'active' ? 'suppliers' : `${data.count} ${type === 'payables' ? 'invoices' : type === 'paid' ? 'payments' : 'invoices'}`;

		// Add exchange rate info for non-EGP currencies
		let exchangeRateInfo = '';
		if (currency !== 'EGP' && window.dashboard_exchange_rates && window.dashboard_exchange_rates[currency]) {
			const rate = parseFloat(window.dashboard_exchange_rates[currency]).toFixed(2);
			exchangeRateInfo = `<div class="card-exchange-rate">1 ${currency} = ${rate} EGP</div>`;
		}

		return `
			<div class="category-card-new" style="border-left: 4px solid ${color}">
				<div class="card-header-new">${title}</div>
				<div class="card-value-new">${value}</div>
				<div class="card-stats-new">
					<span>${subtitle}</span>
				</div>
				${exchangeRateInfo}
				${detailsHtml}
			</div>
		`;
	}

	render_total_card(title, total, currency, color, exchange_rates, breakdown) {
		// Store exchange rates globally for use in render_card
		window.dashboard_exchange_rates = exchange_rates;

		// Show exchange rates
		let ratesHtml = '';
		if (exchange_rates) {
			const rates = [];
			for (const [curr, rate] of Object.entries(exchange_rates)) {
				if (curr !== 'EGP') {
					rates.push(`<span>${curr}: ${parseFloat(rate).toFixed(2)} EGP</span>`);
				}
			}
			if (rates.length > 0) {
				ratesHtml = `<div class="exchange-rates">${rates.join('')}</div>`;
			}
		}

		// Show breakdown by currency
		let breakdownHtml = '';
		if (breakdown && Object.keys(breakdown).length > 0) {
			const items = [];
			for (const [curr, amount] of Object.entries(breakdown)) {
				items.push(`<div class="breakdown-item"><span class="breakdown-currency">${curr}</span><span class="breakdown-amount">${this.format_currency(amount, curr)}</span></div>`);
			}
			breakdownHtml = `<div class="currency-breakdown">${items.join('')}</div>`;
		}

		return `
			<div class="total-card" style="border-left: 4px solid ${color}">
				<div class="card-header-new">${title}</div>
				<div class="card-value-new">${this.format_currency(total, currency)}</div>
				${breakdownHtml}
				${ratesHtml}
			</div>
		`;
	}

	render_top_suppliers(suppliers, exchange_rates) {
		if (!suppliers || suppliers.length === 0) {
			return '<p class="no-data">No supplier data available</p>';
		}

		// Sort by total purchase in EGP
		const suppliersWithEGP = suppliers.map(sup => {
			const rate = exchange_rates[sup.currency] || 1.0;
			return {
				...sup,
				total_purchase_egp: sup.total_purchase_egp || (sup.total_purchase * rate),
				outstanding_egp: sup.outstanding * rate
			};
		});

		suppliersWithEGP.sort((a, b) => b.total_purchase_egp - a.total_purchase_egp);

		// Helper function to calculate relationship duration
		const getDuration = (firstDate) => {
			if (!firstDate) return '';
			const first = new Date(firstDate);
			const now = new Date();
			const diffTime = Math.abs(now - first);
			const diffDays = Math.ceil(diffTime / (1000 * 60 * 60 * 24));

			const years = Math.floor(diffDays / 365);
			const months = Math.floor((diffDays % 365) / 30);

			if (years > 0 && months > 0) {
				return `${years}y ${months}m`;
			} else if (years > 0) {
				return `${years} year${years > 1 ? 's' : ''}`;
			} else if (months > 0) {
				return `${months} month${months > 1 ? 's' : ''}`;
			} else {
				return `${diffDays} days`;
			}
		};

		return `
			<div class="top-suppliers-list">
				${suppliersWithEGP.slice(0, 10).map((sup, index) => `
					<div class="top-supplier-item clickable" onclick="frappe.set_route('Form', 'Supplier', '${sup.supplier}')" title="View supplier details">
						<div class="supplier-rank">#${index + 1}</div>
						<div class="supplier-info">
							<div class="supplier-name">${sup.supplier_name}</div>
							<div class="supplier-stats">
								<span class="stat-item">Total: ${this.format_currency(sup.total_purchase, sup.currency || 'EGP')} (${this.format_currency(sup.total_purchase_egp, 'EGP')})</span>
								<span class="stat-item">${sup.invoice_count} invoices</span>
								<span class="stat-item">📅 ${getDuration(sup.first_invoice_date)}</span>
								<span class="stat-item">Outstanding: ${this.format_currency(sup.outstanding, sup.currency || 'EGP')}</span>
							</div>
						</div>
					</div>
				`).join('')}
			</div>
		`;
	}

	format_currency(value, currency = 'EGP') {
		return new Intl.NumberFormat('en-US', {
			style: 'currency',
			currency: currency,
			minimumFractionDigits: 2,
			maximumFractionDigits: 2
		}).format(value || 0);
	}
}
